### PR TITLE
Improve performance

### PR DESCRIPTION
**Pull request by natano**
- Switch from socket.io to plain websockets to reduce CPU load in the browser as well as on server side
- Don't check canvas dimensions on every frame, but use the window resize event instead
- Move (potentially) expensive computations out of the render loop
- Cap frame rate at 60 FPS. That should prevent excessive CPU usage with high monitor refresh rates (e.g. my 144 Hz monitor at home)

### CODE REVIEW

File: webui/server/server.py [hunk 1]
```diff
@@ -1,52 +1,36 @@
...
-print(' * WebUI can be found at: http://' + ip_address + ':5000')
-
 # Used for developmental purposes. Set this to true when you just want to
 # emulate the serial device instead of actually connecting to one.
-NO_SERIAL = False
-
-def ThreadIsAlive(thread):
-  def has_method(o, name):
-    return callable(getattr(o, name, None))
+NO_SERIAL = True
```
Comment (teejusb) [inline webui/server/server.py:32]: Default this to False

File: webui/server/server.py [hunk 2]
```diff
@@ -285,85 +268,178 @@ def Write(self):
...
+
+async def get_index(request):
+  return web.FileResponse(os.path.join(build_dir, 'index.html'))
+
+
+app = web.Application()
+app.add_routes([
+  web.get('/defaults', get_defaults),
+  web.get('/ws', get_ws),
+  web.get('/', get_index),
```
Comment (teejusb) [inline webui/server/server.py:434]: For previous behavior: '/' shouldn't be added if NO_SERIAL is true (because of the node server separation).

Comment (teejusb) [reply webui/server/server.py:434]: although I'm sure the difference between .get() and the .static() call below

File: webui/src/App.js [hunk 3]
```diff
@@ -145,22 +169,33 @@ function ValueMonitor(props) {
...
-        }
+
+      canvas.width = canvas.clientWidth * dpi;
+      canvas.height = canvas.clientHeight * dpi;
+    };
+
+    setDimensions();
+    window.addEventListener('resize', setDimensions);
+
+    // Add background fill.
```
Comment (teejusb) [inline webui/src/App.js:183]: This comment should be moved below to the createLinearGradient call

File: webui/src/App.js [hunk 4]
```diff
@@ -145,22 +169,33 @@ function ValueMonitor(props) {
...
+
+    const bodyFontFamily = window.getComputedStyle(document.body).getPropertyValue("font-family");
+    const valueLabel = valueLabelRef.current;
+    const thresholdLabel = thresholdLabelRef.current;
+
+    // cap animation to 60 FPS (with slight leeway because monitor refresh rates are not exact)
+    const minFrameDuration = 1000 / 60.1;
```
Comment (teejusb) [inline webui/src/App.js:190]: Can we mention that these durations/timestamps are in ms? Alternatively, rename to minFrameDurationMs or similar.

File: webui/src/App.js [hunk 5]
```diff
@@ -204,10 +236,8 @@ function ValueMonitor(props) {
       ctx.fillRect(0, threshold_pos-Math.floor(threshold_height/2), canvas.width, threshold_height);
 
       // Threshold Label
-      const thresholdLabel = thresholdLabelRef.current;
-      thresholdLabel.innerHTML = kCurThresholds[index];
+      thresholdLabel.innerText = kCurThresholds[index];
       // This is default React CSS font style.
```
Comment (teejusb) [inline webui/src/App.js:240]: Move this comment above to where bodyFontFamily is defined

Comment (natano) [reply webui/src/App.js:240]: done

File: webui/src/App.js (hunk 4 above)
Comment (natano) [reply webui/src/App.js:190]: renamed to minFrameDurationMs

File: webui/src/App.js (hunk 3 above)
Comment (natano) [reply webui/src/App.js:183]: done

File: webui/server/server.py (hunk 2 above)
Comment (natano) [reply webui/server/server.py:434]: `web.get('/', get_index)` - return build/index.html when `/` is requested
`web.get('/plot', get_index)` - return build/index.html when '/' is requested
`web.static('/', build_dir)` - serve all static files in the build folder: e.g.  requesting '/static/css/2.d9ad5f5c.chunk.css' will serve `build/static/css/2.d9ad5f5c.chunk.css`.

File: webui/server/server.py (hunk 1 above)
Comment (natano) [reply webui/server/server.py:32]: done

**Comment by teejusb** [comment]
Can we also update the README.md to indicate we need Python 3.7+ instead of 3.6+ (because of asyncio)

**Pull request merged**